Chartjs attempt

allanjoseph98 opened the pull request:
WIP. Needs a lot of cleaning. I've tried to keep the look and function as similar to the current one as possible.

- [x] Rating distribution
- [x] ACPL chart
- [x] Move-time chart
- [ ] Lag gauge chart
- [ ] Rating history
- [x] Insights

ornicar commented:
Looking promising. We don't need to port all the charts right away. I'd prefer to see one or two charts being completed, which we can merge. Then we attack the next ones later on.

What's missing to ship, say, the ACPL and movetime charts?

ornicar commented:
And what's this about?

https://github.com/lichess-org/lila/pull/13897/files#diff-bbb95265eaa4f7769777a0a9b052e19c617bb3df561c89df2c85785486d5df59R43

allanjoseph98 commented:
Functionally, it should all work. I'm mostly not happy with my code quality. There are lots of improvements to be made.

Problems:
1. I'm not happy about bundle sizes. Currently, chartjs is being bundled custom for every chart; users have to download chartjs in multiple forms instead of using a big fat version that can be cached (like it is with highcharts currently). Overall bundle sizes are approximately the same as before so no real wins or losses there.
2. Annotations on the graph do not exist because I've opted to not use the extra annotations plugin as that adds an extra 30kb to minified file sizes. Instead, I've used some creative liberties to add annotations as plot points themselves. They show up in the tooltip on hover.
3. The movetime chart does not take up the entire canvas area because of consequences of (2.) and using a linear x-axis scale for a bar chart. I thought this was the best compromise. Most users shouldn't notice and it only looks slightly weird when moves are > 100.
4. I'm sure there will be some odd bugs that will come up that I have missed.

Wins:
1. I've added types wherever they were missing
2. Tooltips work a lot better when using a rtl language
3. In insights, I've removed the `numeral` module and used `Intl.NumberFormat` instead to save ~30kb.

> And what's this about? 

It was just a note to self to add dynamic theme-changing which was added with [83b866d](https://github.com/lichess-org/lila/pull/13897/commits/83b866d03d8875ed968ceb58fea487e259be390d) along with other misc changes. Insights support added with [850a98f](https://github.com/lichess-org/lila/pull/13897/commits/850a98fbaa578e8c0ffccc86f3fc2023631df8c4)

ornicar commented:
Thanks.

> I'm not happy about bundle sizes

I did notice an increase in the minified size, yes. Because the charts are not on critical pages (like homepage and gameplay), and because we can render the full page layout before loading them, I think it's ok.
I assume you already looked into chartjs treeshaking.

> Currently, chartjs is being bundled custom for every chart; users have to download chartjs in multiple forms instead of using a big fat version that can be cached (like it is with highcharts currently)

As long as we only load one chartjs instance per page, and only after (rest of) the page is rendered and functional, then that's ok by me. 

> Annotations on the graph do not exist because I've opted to not use the extra annotations plugin as that adds an extra 30kb to minified file sizes.

A bit sad as I think it was nice to see directly on the chart. Kinda incredible that it adds 30kb minified JS. What even is in chartjs?

Of course there will be bugs, but that's ok, we'll sort them out.

ornicar commented:
I think the ACPL and movetime charts already look good enough. Could we have bigger and/or more visible hover points, and also while litting up the "christmas tree"?

ornicar commented:
Almost there! Looks like ui/insight/chart could depend on chart/common to reuse the grid/label colors.

schlawg commented:
I don't want to throw a complication in here last minute, but Allan if your efforts were constrained because you were worried about that extra 30k... I assume there's a reason chart.js + annotations plugin can't be loaded (and cached browserside) dynamically as a module?

allanjoseph98 commented:
I haven't looked into it yet but it should be possible in theory. I prioritised replicating the charts first and then looking at optimisations later so I never got around to looking into it. Compared to highcharts, a similar Chartjs + annotations implementation would be an extra ~30+30=60kb. 

Essentially, the only thing missing is the text next to the annotation line. There is also the movetime chart size issue as a result of hacks that I highlighted earlier. 

I didn't think that displaying extra text was worth a 30kb additional download so I never considered the plugin again. But yeah, you make a good point that since it would be cached it might not be too bad.

I leave it up to you guys to decide.

schlawg commented:
Ok. Down the road, we can look into a custom bundle module with only the things we need from chartjs to be shared by our ui modules that use charts. Perhaps you could just indicate in comments any imports or code that bloat concerns stopped you from adding for future reference.

ornicar commented:
This is great, thank you so much. I'm merging this right now and will deploy it in the days to come, so we can get user feedback.
Feel free to keep polishing it and maybe even get rid of highcharts completely <3
